vendor/models/invoice.py

```diff
@@ -190,6 +197,77 @@ def empty_cart(self):
         for offer in offers:
             self.remove_offer(offer)
 
+    def get_next_billing_date(self):
+        """
+        Return the next billing date, if an invoice has two different billing dates it will return
+        the upcoming one.
+        """
+        recurring_offers = self.order_items.filter(offer__terms__lt=TermType.PERPETUAL)
+        if not recurring_offers.count():
+            return None
+        next_billing_dates = [order_item.offer.get_next_billing_date() for order_item in recurring_offers]
+
+        next_billing_dates.sort()
+
+        return next_billing_dates[0]
+
+    def get_next_billing_price(self):
+        """
+        Returns the price corresponding to the upcoming billing date.
+        """
+        recurring_offers = self.order_items.filter(offer__terms__lt=TermType.PERPETUAL)
+        if not recurring_offers.count():
+            return None
+
+        if recurring_offers.count() == 1:
+            return recurring_offers.first().total
+
+        next_billing_date = recurring_offers.first().offer.get_next_billing_date()
+        next_billing_date_price = recurring_offers.first().total
+
+        for order_item in recurring_offers:
+            if order_item.offer.get_next_billing_date() < next_billing_date:
+                next_billing_date = order_item.offer.get_next_billing_date()
+                next_billing_date_price = order_item.total
+
+        return next_billing_date_price
+
+    def get_savings(self):
+        savings = 0
+
+        savings = self.calculate_subtotal() - self.get_discounts()
+
+        return savings
+
+    def get_discounts(self):
+        """
+        Returns the sum of discounts and trial_discounts. Discounts are related to the offer.price and the offer.product.msrp,
+        while trial discounts are related to the set offer.meta.trial_amount if it has a trial_occurrence.
+        """
+        if 'discounts' in self.vendor_notes:
+            return self.vendor_notes['discounts']
+
+        discounts = 0
+
+        discounts = sum([order_item.discounts for order_item in self.order_items.all() if not self.profile.has_owned_product(order_item.offer.products.all())])
+
+        trial_discounts = sum([order_item.price - order_item.trial_amount for order_item in self.order_items.all() if order_item.offer.has_trial_occurrences()])
+
+        return discounts + trial_discounts
+
+    def save_discounts_vendor_notes(self):
+        """
+        Once an invoice has been completed, this method saves the discounts applied to the invoice
+        in the vendor_notes field. This makes it for faster lookup on get_discounts for future invoice
+        views.
+        """
+        if not isinstance(self.vendor_notes, dict):
+            self.vendor_notes = dict()
+            self.save()
```

Comment:
do you need to save here? you're also saving on line 269

Reply:
first save removed. Thanks :)